### ISSUE TITLE
Read loop hangs after server stop

### ISSUE THREAD

**Issue by bastiandoetsch** [creachadair/jrpc2#108]
We have encountered the situation, that our go executable does not return from WaitStatus.

Apparently, the goroutine to receive messages on the channel does not terminate after the server.stop, and thus the Wait does not return.

What kind of additional information would you need to analyse this in depth?

You can reproduce this running the `snyk-ls` executable, available at https://github.com/snyk/snyk-ls (we do that from an e2e test, located [here](https://github.com/snyk/cli/blob/168ea03070740eb1a2214acef6f9b94065ecfa4e/test/jest/acceptance/language-server-extension.spec.ts)).

Please let me know if I can assist further.

**Comment by creachadair** [comment]
Thanks for the report. From my initial skim of the code, it looks as if there is a handler that is trying to call Stop on the server, and my guess is that is causing a deadlock.

In general it won't be safe for a handler to stop its own server: while Stop holds the internal mutex to clean up the server state, it will exclude the reader from acquiring it while it shuts down. That in turn will cause the wait logic to never complete.

If you'd like a handler to stop its server, you'll probably want to have it signal that a shutdown was requested, and do it from another goroutine after the handler exits.

**Comment by creachadair** [comment]
I poked at this a bit more, and another thing I noticed is that this language server appears to be using stdio to communicate (that is normal). Here, however, instead of the client closing its channel to the server, it sends the server a request to shut itself down. 

When the server stops, it calls `Close` on its channel, which closes its write side (in this case stdout), but it does not have access to close its read side (in this case stdin). 

If the client ignores this and leaves its write side open, the server will continue to wait, which is I think what you're seeing. Probably the simplest way to address this in your integration test is to either (a) make the client stop when it gets EOF from the server, or (b) make the shutdown process on the server side explicitly close its own stdin. E.g., perhaps adding [here](https://github.com/snyk/snyk-ls/blob/39fc0f68d431409f66b733aa20396a05052cff72/application/server/server.go#L410C6-L410C11):

```go
os.Stdin.Close()
```

I see from this that there is a documentation bug for `Channel.Close`, which asserts no further records may be sent _or received_. The latter is not (always) true, and I should fix that.

**Comment by bastiandoetsch** [comment]
Sorry for the late reply - I'll try out closing stdin and see what happens :).

**Comment by bastiandoetsch** [comment]
Closing stdin unfortunately doesn't fix the issue. Tried stopping the server within a go routine outside of the handler, but it still hangs.
```
func exit(srv *jrpc2.Server, c *config.Config) jrpc2.Handler {
	return handler.New(func(_ context.Context) (any, error) {
		logger := c.Logger().With().Str("method", "Exit").Logger()
		logger.Info().Msg("ENTERING")
		logger.Info().Msg("Flushing error reporting...")
		di.ErrorReporter().FlushErrorReporting()
		go func() {
			// wait for the handler to quit, then stop outside of handler the server
			time.Sleep(time.Second)
			logger.Info().Msg("Stopping server...")
			_ = os.Stdin.Close()
			srv.Stop()
		}()
		return nil, nil
	})
}
```

**Comment by bastiandoetsch** [comment]
Yeah, we removed shutdown and exit, as it was not working. But I tested with shutdown and exit and the process didn't quit. Locally the end of the test looks like this:
```
for (let i = 0; i < 45; i++) {
      console.debug('Waiting for diagnostics...');
      if (diagnosticCount > 0) {
        break;
      }
      await sleep(1000);
    }
    await connection.sendRequest('shutdown');
    await connection.sendNotification('exit');
    // cli.kill(9);
```

**Comment by creachadair** [comment]
I attempted to build a minimal repro but it does not in fact reproduce what you're seeing. But perhaps it will give you a basis to isolate the conditions that are different in a way we can turn into a test?

To use:
```shell
unzip repro.sh
cd repro
make
./tcli
```

[repro.zip](https://github.com/creachadair/jrpc2/files/12653308/repro.zip)

**Comment by Simon-Hayden-Dev** [comment]
Hi, I know this is an old issue, but I was playing around and stumbled into the same situation. The issue stems from creating a server, that read from Stdin and writes to Stdout.

Here is my broken minimal setup:

<details>
<summary>
  Dummy JRPC server listening on STDIN
</summary>


```go
package main

import (
	"context"
	"fmt"
	"log"
	"os"
	"os/signal"
	"sync"

	"github.com/creachadair/jrpc2"
	"github.com/creachadair/jrpc2/channel"
	"github.com/creachadair/jrpc2/handler"
)

func main() {
	// Create a channel for os.Stdin + os.Stdout
	ch := channel.Line(os.Stdin, os.Stdout)

	ctx, cancel := context.WithCancelCause(context.Background())
	ctx, _ = signal.NotifyContext(ctx, os.Interrupt)

	var processes sync.WaitGroup

	// Dummy server without any methods
	jrcpServer := jrpc2.NewServer(handler.Map{}, &jrpc2.ServerOptions{
		AllowPush:  true,
		NewContext: func() context.Context { return ctx },
	})
	jrcpServer.Start(ch)
	processes.Go(func() {
		// Terminate the server when the context is stopped
		go func() {
			<-ctx.Done()
			log.Printf("Terminating JRPC server")
			jrcpServer.Stop()
		}()

		// ⚠️ This never returns after interrupt ⚠️
		err := jrcpServer.Wait()
		select {
		case <-ctx.Done():
			return
		default:
		}
		if err != nil && err != ctx.Err() {
			cancel(fmt.Errorf("JRPC server connected to stdio stopped unexpectedly: %w", err))
			log.Printf("JRPC server stopped unexpectedly: %v", err)
		} else {
			cancel(nil)
		}
		log.Printf("Terminated JRPC server")
	})

	// Spawn goroutine to explicitly close Stdin (doesn't work)
	go func() {
		<-ctx.Done()
		log.Println("Closing stdin")
		if err := os.Stdin.Close(); err != nil {
			log.Printf("Failed to close stdin: %v", err)
		}
	}()

	log.Println("Server started")
	log.Println("Pressing Ctrl+C won't do anything. Ctrl+D works.")
	processes.Wait()
	log.Println("Shutting down")
}

```


</details>

I do think, however, that this might be an issue with Go/Linux? Like, when I *explicitly*  call `os.Stdin.Close()`, it *should* emit EOF, but it doesn't. Presumably, this is why the read loop doesn't terminate. But still, if I call `jrcpServer.Stop()`, I would expect that reading from the channel/the underlying `io.Reader` would instantly stop and return.

Disclaimer: I'm new to Go, so I probably don't know what I'm talking about 😅

**Comment by creachadair** [comment]
> I do think, however, that this might be an issue with Go/Linux? Like, when I _explicitly_ call `os.Stdin.Close()`, it _should_ emit EOF, but it doesn't. Presumably, this is why the read loop doesn't terminate. But still, if I call `jrcpServer.Stop()`, I would expect that reading from the channel/the underlying `io.Reader` would instantly stop and return.

In general, a process cannot close its own stdin and expect it to deliver EOF to readers in the same process. You can close the descriptor, and that _may_ cause future reads to deliver EBADF, but with the way the Go runtime polls for activity on descriptors, a pending read may not wake up if a close occurs while the read is already ongoing.

That behaviour is not super-specific to Linux or Go, though the exact outcomes may vary depending on which system you're using (Darwin, for example, does the same thing). Anything that reads from stdin will be affected by this, e.g.,

```go
func main() {
	time.AfterFunc(5*time.Second, func() {
		log.Printf("timeout, closing os.Stdin")
		os.Stdin.Close()
	})

	data, err := io.ReadAll(os.Stdin)
	log.Println(err)
	fmt.Printf("%q\n", data)
}
```

Even after the close, the read will not unblock till the upstream process (e.g., the shell) closes its pipe to that process (say, the user types ctrl-D), or terminates the process some other way. 

If you need the process to be able to control when the input stream closes, you might consider using a Unix-domain socket instead of stdin.

What I'd probably do in this case is stop the server, give it some reasonable lame-duck grace period to send errors and such, then exit the process. Something like:

```go
ctx, cancel := signal.NotifyContext(context.Background(), syscall.SIGINT, syscall.SIGTERM)
defer cancel()
go func() {
   <-ctx.Done()
   log.Printf("signal received, shutting down...")
   srv.Stop()
   time.AfterFunc(5*time.Second, func() { log.Fatal("exiting") })
}()

// ...

srv.Wait()
```

or words to that effect. In other words: After getting the signal to shut down, stop the server and wait "a while" (if you want) to let it terminate its pending queue, and send responses the client may be interested in. But don't block indefinitely expecting the server to exit gracefully unless the client does so—just exit.

Alternatively, depending how your processes are set up, you can signal to the upstream process separately that it should go away. E.g., maybe post a notification to the client that the server is shutting down so it can choose to close on its own. But since the server _is_ shutting down, you'll still probably want to set a timeout and exit anyway if it doesn't take action.

**Comment by Simon-Hayden-Dev** [comment]
Hey, thanks for you fast response and sorry for my late one 😄 

In the end, I did add my own wrapper around `io.Reader` to make it *also* cancelable (with the help of some AI, though it did mess up a bit, which meant manual intervention...). Here is the code.

<details>
<summary>Cancelable Reader</summary>

```go
package iox

import (
	"context"
	"io"
	"sync"
)

type data struct {
	err error
	buf []byte
}

type cancelableReader struct {
	reader        io.Reader
	ctx           context.Context
	nextBuf       chan []byte
	data          chan data
	leftover      []byte
	leftoverError error

	startOnce sync.Once
}

func (r *cancelableReader) Read(b []byte) (int, error) {
	r.startOnce.Do(r.start)

	if len(r.leftover) > 0 {
		n := copy(b, r.leftover)
		r.leftover = r.leftover[n:]
		var err error
		if len(r.leftover) == 0 {
			// No more leftover data, emit the error from the last read
			err = r.leftoverError
		}
		return n, err
	}

	r.nextBuf <- b

	select {
	case <-r.ctx.Done():
		return 0, io.EOF
	case data := <-r.data:
		if len(data.buf) == 0 {
			// No data received
			return 0, data.err
		}
		n := copy(b, data.buf)
		if n == len(data.buf) {
			// All data copied over
			return n, data.err
		}
		r.leftover = data.buf[n:]
		r.leftoverError = data.err
		return n, nil
	}
}

func (r *cancelableReader) start() {
	go func() {
		for {
			if err := r.ctx.Err(); err != nil {
				r.data <- data{err, []byte{}}
				close(r.nextBuf)
				close(r.data)
				return
			}

			buff := <-r.nextBuf

			n, err := r.reader.Read(buff)

			r.data <- data{err, buff[:n]}
		}
	}()
}

func WithContext(ctx context.Context, reader io.Reader) io.Reader {
	return &cancelableReader{
		reader:   reader,
		ctx:      ctx,
		nextBuf:  make(chan []byte),
		data:     make(chan data),
		leftover: []byte{},
	}
}
```

</details>

The implementation leaks a go-routine, but for my use-case, this is fine. It may also not be 100% thread-safe. But hey, I'm just a beginner and it gets the job done, right? 😄
